src/Course/List.hs

```diff
@@ -390,31 +528,39 @@ seqOptional =
  > find (const True) infinity
  Full 0
 -}
+-- TODO: Doesn't work in the infinite case, can we use fold?
+-- find :: (a -> Bool) -> List a -> Optional a
+-- find f xs = foldLeft step init list
+--   where
+--     step = \acc next -> acc <+> if (f next) then Full next else Empty
+--     init = Empty
+--     list = xs
 find :: (a -> Bool) -> List a -> Optional a
-find =
-    error "todo: Course.List#find"
+find _ Nil = Empty
+find f (x :. xs) = if (f x) then Full x else find f xs
 
 {- | Determine if the length of the given list is greater than 4.
 
- > lengthGT4 (1 :. 3 :. 5 :. Nil)
+-- > lengthGT4 (1 :. 3 :. 5 :. 6 :. Nil)
  False
 
- > lengthGT4 Nil
+-- > lengthGT4 Nil
  False
 
- > lengthGT4 (1 :. 2 :. 3 :. 4 :. 5 :. Nil)
+-- > lengthGT4 (1 :. 2 :. 3 :. 4 :. 5 :. Nil)
  True
 
- > lengthGT4 infinity
+-- > lengthGT4 infinity
  True
 -}
 lengthGT4 :: List a -> Bool
-lengthGT4 =
-    error "todo: Course.List#lengthGT4"
+lengthGT4 Nil = False
+lengthGT4 (_ :. _ :. _ :. _ :. _ :. _) = True
+lengthGT4 (_ :. xs) = lengthGT4 xs
```

Comment:
good. this is a situation where `length list > 4` will traverse the whole list.

incidentally, we could do this

```haskell
lengthGT :: Int -> [a] -> Bool
lengthGT n xs = length (zip [0..n] xs) > n
```

Now, the runtime is `min(n, length xs)`

Reply:
an easier way (and one which uses less memory than the `zip` version) just uses `take` :-p

```haskell
lengthGT :: Int -> [a] -> Bool
lengthGT n xs = length (take (n + 1) xs) > n
```

Reply:
Ooh, nice, I wonder how `take` does this under the hood as it seems like it must be doing something similar.